### DISCUSSION TITLE
Can I add an external file to a package when executing cargo lambda build?

### DISCUSSION THREAD

**Discussion by saito-kosuke** [cargo-lambda/cargo-lambda#369] [Q&A]
I'd like to use a dictionary in a lambda when running morphological analysis.
Can I add an external file to a package when executing cargo lambda build?

**Comment by davehorner** [comment]
I would also like to package a file in the deployment zip.  I tried:
```
cargo lambda build --release --arm64 --output-format zip
7za a -tzip target/lambda/myfunc/bootstrap.zip data.db
cargo lambda deploy --enable-function-url --iam-role arn:aws:iam::XXX:role/rolename
```

The file is added to the zip correctly; but for some reason lambda deploy changes the zip file to just contain the bootstrap.  I'm not sure how to add additional files.  I have also tried copying the files into the target/lambda and the target/lambda/myfunc folders.  The build and deploy commands do not package these additional files.

**Comment by saito-kosuke** [comment]
@davehorner 
I've downloaded the external file using reqwest.
You might as well use EFS to store the external file.

**Reply by davehorner** [reply]
@saito-kosuke 
deploying the file in the zip is basic deployment stuff;  local files are safer and faster IMHO.
I'm not familiar with FTS as an acronym.

it should be as simple as dropping it in the archive.  I'm not sure why deploy is rewriting the zip.
about to include_bytes!("data.db"); lol and write it to disk.  j/k   tucking a file in the zip should work.

**Reply by saito-kosuke** [reply]
@davehorner I'm all for it. I should have said EFS rather than FTS.
https://aws.amazon.com/jp/blogs/compute/using-amazon-efs-for-aws-lambda-in-your-serverless-applications/

**Reply by davehorner** [reply]
I ended up using a lambda layer and just refer to things out of /opt.  works fine.  I did try using include_bytes and writing to /tmp.  This worked but it did not work always.  It would give me a malformed data error...but only sometimes.  I'm not sure about that, but layers is a fine solution that allows you to continue to use the cargo lambda deploy method.  I'd close the ticket given we have solutions.

**Comment by bnusunny** [comment]
Good finding. Lambda layer is the current work around for packaging extra files and work well with cargo lambda.

**Reply by davehorner** [reply]
only problem that comes up is the movement to /opt breaks the paths for local development.  Is there a simple way to tell if I'm running locally to adjust the paths during local development?

**Reply by bnusunny** [reply]
You could check the environment variable `AWS_LAMBDA_RUNTIME_API`. This envar is not empty in actual Lambda execution environment.

**Reply by davehorner** [reply]
Hmm.  When I run via the cargo lambda watch I see that env var set.
```
[src\main.rs:107] env::var("AWS_LAMBDA_RUNTIME_API") = Ok(
    "http://127.0.0.1:9000/.rt/myfunc",
)
```
I can contains a 127.0.0.1, but I figured I'd mention that I'm seeing that non-empty on local invocations.

**Reply by bnusunny** [reply]
Oh, cargo lambda watch sets this envar.

**Reply by davehorner** [reply]
I'm seeing AWS_LAMBDA_RUNTIME_API: 127.0.0.1:9001 is set in my actual lambda running on AWS.  So contains 127.0.0.1 is not sufficient.  contains("/.rt/") works for now but it's a little weird.  It might be worth adding an environment variable in the watch environment. CARGO_LAMBDA or something, could be the version or something.  /.rt/ works.
